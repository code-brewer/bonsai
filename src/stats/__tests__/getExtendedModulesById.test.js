/*
 * @flow
 */

import type {ExtendedModule} from '../../types/Stats';

import {
  defaultModule,
  defaultExtendedModule,
  reasonFromModule,
} from '../../__test_helpers__/defaults';

import getExtendedModulesById, {
  calculateModuleSizes
} from '../getExtendedModulesById';

const ENTRY_ZERO = defaultModule({
  id: 0,
  name: 'Entry Zero',
  size: 100,
  reasons: [],
});

const ENTRY_ONE = defaultModule({
  id: 1,
  name: 'Entry One',
  size: 100,
  reasons: [],
});

const ZERO_A = defaultModule({
  id: 2,
  name: 'Module Zero-A',
  size: 250,
  reasons: [
    reasonFromModule(ENTRY_ZERO),
  ],
});

const COMMON = defaultModule({
  id: 3,
  name: 'Module Common',
  size: 200,
  reasons: [
    reasonFromModule(ENTRY_ZERO),
    reasonFromModule(ENTRY_ONE),
  ],
});

const BIG_MODULE = defaultModule({
  id: 5,
  name: 'Big Module',
  size: 1000,
  reasons: [
    reasonFromModule(ZERO_A),
    reasonFromModule(COMMON),
  ],
});

const ENTRY_ONE_A = defaultModule({
  id: 600,
  name: 'Module One-A',
  size: 600,
  reasons: [
    reasonFromModule(ENTRY_ONE),
  ],
});

const ENTRY_ONE_B = defaultModule({
  id: 300,
  name: 'Module One-B',
  size: 300,
  reasons: [
    reasonFromModule(ENTRY_ONE_A),
  ],
});

const MODULE_COUNT = 7;
const MODULES = [
  ENTRY_ZERO,
  ENTRY_ONE,
  ZERO_A,
  COMMON,
  BIG_MODULE,
  ENTRY_ONE_A,
  ENTRY_ONE_B,
];

describe('getExtendedModulesById', () => {
  it('should extend all the modules', () => {
    const modulesById = getResults(MODULES);

    expect({
      ...modulesById[ENTRY_ZERO.id],
      cumulativeSize: 0, // reset value
      requirements: [], // reset value
      requirementsCount: 0, // reset value
    }).toEqual(defaultExtendedModule(ENTRY_ZERO));
    expect(modulesById).toMatchSnapshot();
  });

  it('should have the right fixture sizes', () => {
    expect(MODULES.map((module) => module.size)).toEqual([
      100, 100, 250, 200, 1000, 600, 300,
    ]);
  });

  it('should have the right cumulative sizes', () => {
    const modulesById = getResults(MODULES);

    expect(modulesById[ENTRY_ZERO.id].cumulativeSize).toBe(1200);
    expect(modulesById[ENTRY_ONE.id].cumulativeSize).toBe(1350);
    expect(modulesById[ZERO_A.id].cumulativeSize).toBe(750);
    expect(modulesById[COMMON.id].cumulativeSize).toBe(700);
    expect(modulesById[BIG_MODULE.id].cumulativeSize).toBe(1000);
    expect(modulesById[ENTRY_ONE_A.id].cumulativeSize).toBe(900);
    expect(modulesById[ENTRY_ONE_B.id].cumulativeSize).toBe(300);
  });

  it('should return the `size` as cumulative for modules without requirements', () => {
    const modulesById = getResults(MODULES);

    const modulesWithoutRequirements = [ENTRY_ONE_B.id, BIG_MODULE.id];

    modulesWithoutRequirements.forEach((id) => {
      expect(modulesById[id].size).toBeGreaterThan(0);
      expect(modulesById[id].cumulativeSize).toBe(modulesById[id].size);
    });
  });

  it('the sum of the extended modules should make sense', () => {
    const modulesById = getResults(MODULES);

    // $FlowFixMe: flow thinks `values` returns `Array<mixed>`
    const resultValues: Array<ExtendedModule> = Object.values(modulesById);

    const sumSizes = resultValues.reduce((sum, eModule) => {
      return sum + eModule.size;
    }, 0);

    const nothingRequires = resultValues.filter((eModule) => {
      return eModule.requiredByCount === 0;
    });
    const sumCumulativeTopLevelSizes = nothingRequires.reduce((sum, eModule) => {
      return sum + eModule.cumulativeSize;
    }, 0);

    const SUM_OF_ALL_MODULES = 2550;
    expect(sumSizes).toBeCloseTo(SUM_OF_ALL_MODULES);
    expect(sumCumulativeTopLevelSizes).toBe(SUM_OF_ALL_MODULES);
    expect(sumSizes).toBe(sumCumulativeTopLevelSizes);
  });

  it('should extend all the trimmed modules', () => {
    const modulesById = getResults(MODULES);

    // $FlowFixMe: flow thinks `values` returns `Array<mixed>`
    const resultValues: Array<ExtendedModule> = Object.values(modulesById);

    const trimResults = resultValues.map((module: ExtendedModule) => {
      return {
        id: module.id,
        identifier: module.identifier,
        size: module.size,
        cumulativeSize: module.cumulativeSize,
        requiredByCount: module.requiredByCount,
        requirements: module.requirements.map((requirement) => {
          const rModule = modulesById[requirement.id];
          return {
            id: rModule.id,
            identifier: rModule.identifier,
            size: rModule.size,
            cumulativeSize: rModule.cumulativeSize,
            requiredByCount: rModule.requiredByCount,
            contributingSize: rModule.cumulativeSize / rModule.requiredByCount,
          };
        }),
      };
    });

    expect(trimResults).toMatchSnapshot();
  });
});

function getResults(modules) {
  const modulesById = calculateModuleSizes(getExtendedModulesById(modules));
  if (!modulesById) {
    throw Error('getExtendedModulesById returned null which was not expected');
  }

  expect(modulesById).not.toBeNull();
  expect(Object.keys(modulesById)).toHaveLength(MODULE_COUNT);

  return modulesById;
}
