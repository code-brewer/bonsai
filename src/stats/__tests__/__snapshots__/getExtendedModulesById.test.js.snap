// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExtendedModulesById should extend all the modules 1`] = `
Object {
  "0": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 1200,
    "errors": 0,
    "failed": false,
    "id": 0,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Entry Zero",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [],
    "requiredBy": Array [],
    "requiredByCount": 0,
    "requirements": Array [
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 2,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Module Zero-A",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Entry Zero",
            "moduleId": 0,
            "moduleIdentifier": "",
            "moduleName": "Entry Zero",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 250,
        "usedExports": true,
        "warnings": 0,
      },
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 3,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Module Common",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Entry Zero",
            "moduleId": 0,
            "moduleIdentifier": "",
            "moduleName": "Entry Zero",
            "type": "",
            "userRequest": "",
          },
          Object {
            "loc": "",
            "module": "Entry One",
            "moduleId": 1,
            "moduleIdentifier": "",
            "moduleName": "Entry One",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 200,
        "usedExports": true,
        "warnings": 0,
      },
    ],
    "requirementsCount": 2,
    "size": 100,
    "usedExports": true,
    "warnings": 0,
  },
  "1": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 1350,
    "errors": 0,
    "failed": false,
    "id": 1,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Entry One",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [],
    "requiredBy": Array [],
    "requiredByCount": 0,
    "requirements": Array [
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 3,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Module Common",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Entry Zero",
            "moduleId": 0,
            "moduleIdentifier": "",
            "moduleName": "Entry Zero",
            "type": "",
            "userRequest": "",
          },
          Object {
            "loc": "",
            "module": "Entry One",
            "moduleId": 1,
            "moduleIdentifier": "",
            "moduleName": "Entry One",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 200,
        "usedExports": true,
        "warnings": 0,
      },
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 600,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Module One-A",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Entry One",
            "moduleId": 1,
            "moduleIdentifier": "",
            "moduleName": "Entry One",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 600,
        "usedExports": true,
        "warnings": 0,
      },
    ],
    "requirementsCount": 2,
    "size": 100,
    "usedExports": true,
    "warnings": 0,
  },
  "2": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 750,
    "errors": 0,
    "failed": false,
    "id": 2,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Module Zero-A",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [
      Object {
        "loc": "",
        "module": "Entry Zero",
        "moduleId": 0,
        "moduleIdentifier": "",
        "moduleName": "Entry Zero",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredBy": Array [
      Object {
        "loc": "",
        "module": "Entry Zero",
        "moduleId": 0,
        "moduleIdentifier": "",
        "moduleName": "Entry Zero",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredByCount": 1,
    "requirements": Array [
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 5,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Big Module",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Module Zero-A",
            "moduleId": 2,
            "moduleIdentifier": "",
            "moduleName": "Module Zero-A",
            "type": "",
            "userRequest": "",
          },
          Object {
            "loc": "",
            "module": "Module Common",
            "moduleId": 3,
            "moduleIdentifier": "",
            "moduleName": "Module Common",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 1000,
        "usedExports": true,
        "warnings": 0,
      },
    ],
    "requirementsCount": 1,
    "size": 250,
    "usedExports": true,
    "warnings": 0,
  },
  "3": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 700,
    "errors": 0,
    "failed": false,
    "id": 3,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Module Common",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [
      Object {
        "loc": "",
        "module": "Entry Zero",
        "moduleId": 0,
        "moduleIdentifier": "",
        "moduleName": "Entry Zero",
        "type": "",
        "userRequest": "",
      },
      Object {
        "loc": "",
        "module": "Entry One",
        "moduleId": 1,
        "moduleIdentifier": "",
        "moduleName": "Entry One",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredBy": Array [
      Object {
        "loc": "",
        "module": "Entry Zero",
        "moduleId": 0,
        "moduleIdentifier": "",
        "moduleName": "Entry Zero",
        "type": "",
        "userRequest": "",
      },
      Object {
        "loc": "",
        "module": "Entry One",
        "moduleId": 1,
        "moduleIdentifier": "",
        "moduleName": "Entry One",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredByCount": 2,
    "requirements": Array [
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 5,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Big Module",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Module Zero-A",
            "moduleId": 2,
            "moduleIdentifier": "",
            "moduleName": "Module Zero-A",
            "type": "",
            "userRequest": "",
          },
          Object {
            "loc": "",
            "module": "Module Common",
            "moduleId": 3,
            "moduleIdentifier": "",
            "moduleName": "Module Common",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 1000,
        "usedExports": true,
        "warnings": 0,
      },
    ],
    "requirementsCount": 1,
    "size": 200,
    "usedExports": true,
    "warnings": 0,
  },
  "300": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 300,
    "errors": 0,
    "failed": false,
    "id": 300,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Module One-B",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [
      Object {
        "loc": "",
        "module": "Module One-A",
        "moduleId": 600,
        "moduleIdentifier": "",
        "moduleName": "Module One-A",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredBy": Array [
      Object {
        "loc": "",
        "module": "Module One-A",
        "moduleId": 600,
        "moduleIdentifier": "",
        "moduleName": "Module One-A",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredByCount": 1,
    "requirements": Array [],
    "requirementsCount": 0,
    "size": 300,
    "usedExports": true,
    "warnings": 0,
  },
  "5": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 1000,
    "errors": 0,
    "failed": false,
    "id": 5,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Big Module",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [
      Object {
        "loc": "",
        "module": "Module Zero-A",
        "moduleId": 2,
        "moduleIdentifier": "",
        "moduleName": "Module Zero-A",
        "type": "",
        "userRequest": "",
      },
      Object {
        "loc": "",
        "module": "Module Common",
        "moduleId": 3,
        "moduleIdentifier": "",
        "moduleName": "Module Common",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredBy": Array [
      Object {
        "loc": "",
        "module": "Module Zero-A",
        "moduleId": 2,
        "moduleIdentifier": "",
        "moduleName": "Module Zero-A",
        "type": "",
        "userRequest": "",
      },
      Object {
        "loc": "",
        "module": "Module Common",
        "moduleId": 3,
        "moduleIdentifier": "",
        "moduleName": "Module Common",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredByCount": 2,
    "requirements": Array [],
    "requirementsCount": 0,
    "size": 1000,
    "usedExports": true,
    "warnings": 0,
  },
  "600": Object {
    "assets": Array [],
    "built": false,
    "cacheable": false,
    "chunks": Array [],
    "cumulativeSize": 900,
    "errors": 0,
    "failed": false,
    "id": 600,
    "identifier": "",
    "index": 0,
    "index2": 0,
    "issuer": "",
    "issuerId": "",
    "loops": Array [],
    "name": "Module One-A",
    "optional": false,
    "prefetched": false,
    "providedExports": null,
    "reasons": Array [
      Object {
        "loc": "",
        "module": "Entry One",
        "moduleId": 1,
        "moduleIdentifier": "",
        "moduleName": "Entry One",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredBy": Array [
      Object {
        "loc": "",
        "module": "Entry One",
        "moduleId": 1,
        "moduleIdentifier": "",
        "moduleName": "Entry One",
        "type": "",
        "userRequest": "",
      },
    ],
    "requiredByCount": 1,
    "requirements": Array [
      Object {
        "assets": Array [],
        "built": false,
        "cacheable": false,
        "chunks": Array [],
        "errors": 0,
        "failed": false,
        "id": 300,
        "identifier": "",
        "index": 0,
        "index2": 0,
        "issuer": "",
        "issuerId": "",
        "name": "Module One-B",
        "optional": false,
        "prefetched": false,
        "providedExports": null,
        "reasons": Array [
          Object {
            "loc": "",
            "module": "Module One-A",
            "moduleId": 600,
            "moduleIdentifier": "",
            "moduleName": "Module One-A",
            "type": "",
            "userRequest": "",
          },
        ],
        "size": 300,
        "usedExports": true,
        "warnings": 0,
      },
    ],
    "requirementsCount": 1,
    "size": 600,
    "usedExports": true,
    "warnings": 0,
  },
}
`;

exports[`getExtendedModulesById should extend all the trimmed modules 1`] = `
Array [
  Object {
    "cumulativeSize": 1200,
    "id": 0,
    "identifier": "",
    "requiredByCount": 0,
    "requirements": Array [
      Object {
        "contributingSize": 750,
        "cumulativeSize": 750,
        "id": 2,
        "identifier": "",
        "requiredByCount": 1,
        "size": 250,
      },
      Object {
        "contributingSize": 350,
        "cumulativeSize": 700,
        "id": 3,
        "identifier": "",
        "requiredByCount": 2,
        "size": 200,
      },
    ],
    "size": 100,
  },
  Object {
    "cumulativeSize": 1350,
    "id": 1,
    "identifier": "",
    "requiredByCount": 0,
    "requirements": Array [
      Object {
        "contributingSize": 350,
        "cumulativeSize": 700,
        "id": 3,
        "identifier": "",
        "requiredByCount": 2,
        "size": 200,
      },
      Object {
        "contributingSize": 900,
        "cumulativeSize": 900,
        "id": 600,
        "identifier": "",
        "requiredByCount": 1,
        "size": 600,
      },
    ],
    "size": 100,
  },
  Object {
    "cumulativeSize": 750,
    "id": 2,
    "identifier": "",
    "requiredByCount": 1,
    "requirements": Array [
      Object {
        "contributingSize": 500,
        "cumulativeSize": 1000,
        "id": 5,
        "identifier": "",
        "requiredByCount": 2,
        "size": 1000,
      },
    ],
    "size": 250,
  },
  Object {
    "cumulativeSize": 700,
    "id": 3,
    "identifier": "",
    "requiredByCount": 2,
    "requirements": Array [
      Object {
        "contributingSize": 500,
        "cumulativeSize": 1000,
        "id": 5,
        "identifier": "",
        "requiredByCount": 2,
        "size": 1000,
      },
    ],
    "size": 200,
  },
  Object {
    "cumulativeSize": 1000,
    "id": 5,
    "identifier": "",
    "requiredByCount": 2,
    "requirements": Array [],
    "size": 1000,
  },
  Object {
    "cumulativeSize": 300,
    "id": 300,
    "identifier": "",
    "requiredByCount": 1,
    "requirements": Array [],
    "size": 300,
  },
  Object {
    "cumulativeSize": 900,
    "id": 600,
    "identifier": "",
    "requiredByCount": 1,
    "requirements": Array [
      Object {
        "contributingSize": 300,
        "cumulativeSize": 300,
        "id": 300,
        "identifier": "",
        "requiredByCount": 1,
        "size": 300,
      },
    ],
    "size": 600,
  },
]
`;
